### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by chinedufn**
This PR creates `mdbook-superimport`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+```toml
+# In your book.toml
+[preprocessor.superimport]
+```
+
+```md
+<!-- In your markdown files -->
+
+\`\`\`
```
Comment (murtyjones) [inline README.md]: these escaped backticks are rendering with the slashs. IDK if thats your intent but just letting you know

File: book/src/test-cases/tag-import/fixture.css [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.this-will-not-be-included {
```
Comment (murtyjones) [inline book/src/test-cases/tag-import/fixture.css:1]: for my own understanding, what does `fixture` mean in this context?

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    ) -> Result<Book, mdbook::errors::Error> {
+        debug!("Running `run` method in superimport Preprocessor trait impl");
+
+        let book_src_dir = ctx.root.join(&ctx.config.book.src);
+
+        for section in book.sections.iter_mut() {
+            process_chapter(section, &book_src_dir)?;
+        }
+
+        Ok(book)
```
Comment (murtyjones) [inline src/lib.rs:40]: Again for my own learning, is `Ok` like `assert()`? or is it a callback?

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+                                  # Non escaped import -> `{{ #superimport some-file.txt@some-tag }}`
+    \{\{\s*                         # opening braces and whitespace
+    \#superimport                       # #superimport
+    \s+                             # separating whitespace
+    (?P<file>[a-zA-Z0-9\s_.\-/\\]+) # some-file.txt
+    @                               # @ symbol that denotes the name of a tag
+    (?P<tag>[a-zA-Z0-9_.\-]+)       # some-tag (alphanumeric underscores and dashes)
+    \s*\}\}                         # whitespace and closing braces
+  "
+  ).unwrap();
```
Comment (murtyjones) [inline src/lib.rs:163]: LEGIT.

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+[\n\r]                     # New line right before the end import tag
+
+.*?                        # Characters between start of end import line and end import tag
+
+@superimport
+\s+                        # Separating whitespace
+end
+\s+                        # Separating whitespace
+{tag}
+",
```
Comment (murtyjones) [inline src/lib.rs:257]: pure beauty.

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        let chapter_dir = "book/src/test-cases/tag-import";
+        let chapter_dir = format!("{}/{}", env!("CARGO_MANIFEST_DIR"), chapter_dir);
+
+        let content_between_tags = simport.read_content_between_tags(&chapter_dir.into());
+
+        let expected_content = r#"
+.this-will-be-included {
+  display: block;
+}
+"#;
```
Comment (murtyjones) [inline src/lib.rs:307]: 🍟 This is a pure preference thing but i'm always worried some junior or uninitiated dev coming along and being like "whoa that looks weird let me indent these lines" and ruin the tests. I'd either add a `/// Intentional lack of indenting` or do this:

```
let expected_content = r#"\n.this-will-be-included {\n  display: block;\n}\n"#;
```

or if valid:
```
let expected_content = r#"\n" +
     ".this-will-be-included {\n" +
         "display: block;\n" +
     "}\n"#;
```

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+
+        process_chapter(&mut item, &"".into()).unwrap();
+
+        let expected_content = r#"# Tag Import
+
+```md
+
+.this-will-be-included {
+  display: block;
+}
```
Comment (murtyjones) [inline src/lib.rs:326]: 🍟  same comment on the indenting

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,121 @@
+version: 2
+
+jobs:
+
+  test:
```
Comment (CodyCrowOK) [inline .circleci/config.yml:5]: I successfully ran this job locally with `circleci local execute --job test`

**Review by murtyjones — APPROVED** [review]
Couple small fries and questions mainly for my own learning but looks AMAZING

File: README.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  let also_imported = "Everyting between start/end gets imported."
+  // ...
+
+  // @superimport end some-tag-name-here
+}
+```
+
+## [Full Guide](https://tailwind.github.io/mdbook-superimport/)
+
+[The mdbook-superimport guide](https://tailwind.github.io/mdbook-superimport/)
```
Comment (CodyCrowOK) [inline README.md:54]: I'm assuming this link won't be dead after this is merged?

File: test.sh [hunk 10]
```diff
@@ -0,0 +1 @@
+cargo build --all && cargo test --all
```
Comment (CodyCrowOK) [inline test.sh:3]: Need a shebang for portability

```suggestion
#!/usr/bin/env bash
cargo build --all && cargo test --all
```

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        );
+
+        let path = Path::join(&chapter_dir, &self.file);
+
+        let content = String::from_utf8(::std::fs::read(&path).unwrap()).unwrap();
+
+        // @superimport start foo <--- this line is not captured
+        // ... match all of these
+        // ... lines between the
+        // ... start and end tags <--- this line is not captured
```
Comment (CodyCrowOK) [inline src/lib.rs]: Should the `<--- this line is not captured` be on the next line?

File: book/book.toml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+[book]
+authors = ["Chinedu Francis Nwafili"]
+multilingual = false
+src = "src"
+title = "The Mdbook Superimport Book"
+
+# @superimport start super-section
+[preprocessor.superimport]
```
Comment (CodyCrowOK) [inline book/book.toml:8]: This is how preprocessors get added, right?

**Review by CodyCrowOK — CHANGES_REQUESTED** [review]
Great work. Left a small request but overall this is great.

File: README.md (hunk 9 above)
Comment (chinedufn) [reply README.md:54]: Yup guide is automatically deployed from master builds

File: book/book.toml (hunk 12 above)
Comment (chinedufn) [reply book/book.toml:8]: Yup!

File: book/src/test-cases/tag-import/fixture.css (hunk 2 above)
Comment (chinedufn) [reply book/src/test-cases/tag-import/fixture.css:1]: Something that only exists to be used for a test. 

Commonly used but definitely unclear...

So ... this is a good point. Going forwards I'll use a more clear word.

File: src/lib.rs (hunk 3 above)
Comment (chinedufn) [reply src/lib.rs:40]: For sure

It's neither, it's a [Result](https://doc.rust-lang.org/std/result/)

tl;dr 

```
Result {
  Ok(T) <------ this means that everything worked
  Err      <------ there was an error
}
```

Note on line 31 we're returning  a `Result` which is just an enum with two variants `Result::Ok(T)` and `Result::Err`

Comment (chinedufn) [reply src/lib.rs:40]: So here we're returning `Ok(book)`

File: src/lib.rs (hunk 4 above)
Comment (chinedufn) [reply src/lib.rs:163]: Thanks for the guidance!!

File: src/lib.rs (hunk 6 above)
Comment (chinedufn) [reply src/lib.rs:307]: I guess tests would fail and said dev would see that - but I'm all for making things more clear!

a0c248d

File: test.sh (hunk 10 above)
Comment (chinedufn) [reply test.sh:3]: Thanks! ac28df1

**Comment by chinedufn** [comment]
All feedback addressed thanks!
